### DISCUSSION TITLE
How to convert the body from axum request into "reqwest" body?

### DISCUSSION THREAD

**Discussion by spikeHollow** [tokio-rs/axum#2603] [Q&A] [answered]
### Summary

I'm trying to convert the body received from axum Request as body to reqwest client. But, when I do req.body(), it gives just &Body which doesn't give any info about type of Body.
I'm doing this inside a closure for tower::service_fn. It only works if the closure argument is axum Request. I'm stuck at this point. Any kind of help would be really appreciated.
Ideally, I want the body to be directly streamed from request to the reqwest client, but this needs the ownership of the body.

### axum version

0.7.4

**Comment by davidpdrsn** [comment]
Basically what https://github.com/tokio-rs/axum/blob/main/examples/reqwest-response/src/main.rs does

**Reply by quadruple-output** [reply]
@davidpdrsn Really? As far as I understand the example, it converts a reqwest body into an axum body, which is the opposite direction of what has been asked for. (And, btw., I think the example drops the headers received from reqwest (they are converted, but dropped)).

**Comment by dcadenas** [comment]
You need to `use reqwest::Body;` that has a From for that

**Comment by Threated** [comment, marked as answer]
I usually do this:
```rs
fn convert(req: axum::extract::Request) -> reqwest::Request {
    let req = req.map(|body| reqwest::Body::wrap_stream(body.into_data_stream()));
    reqwest::Request::try_from(req).expect("http::Uri to url::Url conversion failed")
}
```
This only works with the `stream` feature of `reqwest`.
The `reqwest::Request` can then be executed with `reqwest::Client::execute`.
